src/peer_select.cc

```diff
@@ -256,154 +385,106 @@ PeerSelector::CheckAlwaysDirectDone(allow_t answer, void *data)
 }
 
-/// A single DNS resolution loop iteration: Converts selected FwdServer to IPs.
-void
-PeerSelector::resolveSelected()
+static void
+checkLastPeerAccessWrapper(allow_t answer, void *data)
 {
-    if (selectionAborted())
-        return;
+    PeerSelector *selector = static_cast<PeerSelector *>(data);
+    selector->checkLastPeerAccess(answer);
+}
 
+void
+PeerSelector::checkLastPeerAccess(allow_t answer)
+{
     FwdServer *fs = servers;
+    assert(fs);
+    bool peerGone = fs->_peer.raw() && ! fs->_peer.valid();
 
-    // Bug 3243: CVE 2009-0801
-    // Bypass of browser same-origin access control in intercepted communication
-    // To resolve this we must use only the original client destination when going DIRECT
-    // on intercepted traffic which failed Host verification
-    const HttpRequest *req = request;
-    const bool isIntercepted = !req->flags.redirected &&
-                               (req->flags.intercepted || req->flags.interceptTproxy);
-    const bool useOriginalDst = Config.onoff.client_dst_passthru || !req->flags.hostVerified;
-    const bool choseDirect = fs && fs->code == HIER_DIRECT;
-    if (isIntercepted && useOriginalDst && choseDirect) {
-        // check the client is still around before using any of its details
-        if (req->clientConnectionManager.valid()) {
-            // construct a "result" adding the ORIGINAL_DST to the set instead of DIRECT
-            Comm::ConnectionPointer p = new Comm::Connection();
-            p->remote = req->clientConnectionManager->clientConnection->local;
-            fs->code = ORIGINAL_DST; // fs->code is DIRECT. This fixes the display.
-            handlePath(p, *fs);
-        }
-
-        // clear the used fs and continue
-        servers = fs->next;
-        delete fs;
-        resolveSelected();
-        return;
-    }
-
-    // convert the list of FwdServer destinations into destinations IP addresses
-    if (fs && wantsMoreDestinations()) {
-        // send the next one off for DNS lookup.
-        const char *host = fs->_peer.valid() ? fs->_peer->host : request->url.host();
-        debugs(44, 2, "Find IP destination for: " << url() << "' via " << host);
-        Dns::nbgethostbyname(host, this);
+    if (answer.denied() || peerGone) {
+        currentServer = currentServer->next;
+        selectMore();
         return;
     }
 
-    // Bug 3605: clear any extra listed FwdServer destinations, when the options exceeds max_foward_tries.
-    // due to the allocation method of fs, we must deallocate each manually.
-    // TODO: use a std::list so we can get the size and abort adding whenever the selection loops reach Config.forward_max_tries
-    if (fs) {
-        assert(fs == servers);
-        while (fs) {
-            servers = fs->next;
-            delete fs;
-            fs = servers;
-        }
+    if (fs->_peer.valid()) {
+        if (fs->code == CD_PARENT_HIT || fs->code == CD_SIBLING_HIT || fs->code == CLOSEST_PARENT) {
+#if USE_CACHE_DIGESTS
```

Comment:
I recommend moving this whole `#if` block before or after the surrounding `CD_PARENT_HIT || CD_SIBLING_HIT || CLOSEST_PARENT` `if`statement. The CPU cost increase would be negligible, but the code may become a lot more readable.